docs/migration_to_3.0.0.md

```diff
@@ -0,0 +1,80 @@
+# Migrating from Wasmer 2.x to Wasmer 3.0.0
+
+This document will describe the differences between Wasmer 2.x and Wasmer 3.0.0
+and provide examples to make migrating to the new API as simple as possible.
+
+## Table of Contents
+
+- [Rationale for changes in 3.0.0](#rationale-for-changes-in-300)
+- [How to use Wasmer 3.0.0](#how-to-use-wasmer-300)
+  - [Installing Wasmer CLI](#installing-wamser-cli)
+  - [Using Wasmer 3.0.0](#using-wamser-300)
+- [Project structure](#project-structure)
+- [Differences](#differences)
+  - [Managing imports](#managing-imports)
+
+## Rationale for changes in 3.0.0
+
+This version introduces the following changes to make the Wasmer API more ergonomic and safe:
+
+1. `ImportsObject` and the traits `Resolver`, `NamedResolver`, etc have been removed and replaced with a single simple type `Imports`. This reduces the complexity of setting up an `Instance`. The helper macro `imports!` can still be used.
+
+## How to use Wasmer 3.0.0
+
+### Installing Wasmer CLI
+
+See [wasmer.io] for installation instructions.
+
+If you already have wasmer installed, run `wasmer self-update`.
+
+Install the latest versions of Wasmer with [wasmer-nightly] or by following the
+steps described in the documentation: [Getting Started][getting-started].
+
+### Using Wasmer 3.0.0
+
+See the [examples] to find out how to do specific things in Wasmer 3.0.0.
+
+## Project Structure
+
+TODO
+
+## Differences
+
+### Managing imports
+
+Instantiating a Wasm module is similar to 2.x.x.:
+
+```rust
+let import_object: Imports = imports! {
+    "env" => {
+        "host_function" => host_function,
+    },
+};
+let instance = Instance::new(&module, &import_object).expect("Could not instantiate module.");
+```
+
+You can also build the `Imports` object manually:
+
+```rust
+let mut import_object: Imports = Imports::new();
+import_object.define("env", "host_function", host_function);
+let instance = Instance::new(&module, &import_object).expect("Could not instantiate module.");
+```
+
```

Comment:
Before we used to have the `ChainLeftNamedResolver` (or similar, forgot the exact name). We should probably put an example on how to merge two import objects (left to right and right to left)

Reply:
We hadn't talked about that, only that `*Resolver` traits are removed, that's why this wasn't included. This can implemented by using `IntoIterator`:

```rust
let imports1: Imports = todo!();
let mut imports2: Imports = todo!();

for (ns, name, ext) in imports1.into_iter() {
    imports2.define(&ns &name, ext);
}
```

Shall I add this into the documentation?

We can also add `chain_back` and `chain_front` methods for ease, or even implement [`std::ops::Add`](https://doc.rust-lang.org/stable/std/ops/trait.Add.html) and [`std::ops::AddAssign`](https://doc.rust-lang.org/stable/std/ops/trait.AddAssign.html):

```rust
let imports1: Imports = todo!();
let mut imports2: Imports = todo!();

imports2 += imports1;
```

But I don't know if this functionality is worth the extra API.

Reply:
You could implement the `Extend` trait, which would allow you to call `imports.extend(other_imports)` since `Imports` implements `IntoIterator`.